README.md
# MahaLend Protocol

This repository contains the smart contracts source code and markets configuration for the MahaLend Protocol.

- [Documentation](https://docs.mahalend.com)
- [Discord](https://discord.gg/mahadao)
- [Governance Forums](https://discuss.mahadao.com)

## Getting Started

You can install `@mahalend/core` as an NPM package in your Hardhat or Truffle project to import the contracts and interfaces:

`npm install @mahalend/core`

Import at Solidity files:

```
import {IPool} from "@mahalend/core/contracts/interfaces/IPool.sol";

contract Misc {

  function supply(address pool, address token, address user, uint256 amount) public {
    IPool(pool).supply(token, amount, user, 0);
    {...}
  }
}
```

The JSON artifacts with the ABI and Bytecode are also included into the bundled NPM package at `artifacts/` directory.

Import JSON file via Node JS `require`:

```
const PoolV3Artifact = require('@mahalend/core/artifacts/contracts/protocol/pool/Pool.sol/Pool.json');

// Log the ABI into console
console.log(PoolV3Artifact.abi)
```

## Setup

The repository uses Docker Compose to manage sensitive keys and load the configuration. Prior any action like test or deploy, you must run `docker-compose up` to start the `contracts-env` container, and then connect to the container console via `docker-compose exec contracts-env bash`.

Follow the next steps to setup the repository:

- Install `docker` and `docker-compose`
- Create an enviroment file named `.env` and fill the next enviroment variables

```
# Add Alchemy or Infura provider keys, alchemy takes preference at the config level
ALCHEMY_KEY=""
INFURA_KEY=""


# Optional, if you plan to use Tenderly scripts
TENDERLY_PROJECT=""
TENDERLY_USERNAME=""

```

## Test

You can run the full test suite with the following commands:

```
# In one terminal
docker-compose up

# Open another tab or terminal
docker-compose exec contracts-env bash

# A new Bash terminal is prompted, connected to the container
npm run test
```


CHANGELOG.md
# Changelog

## [1.16.2](https://github.com/aave/aave-v3-core/compare/v1.16.1...v1.16.2) (2022-07-28)

### Features

- bump ci node.js to 16 ([82a11d2](https://github.com/aave/aave-v3-core/commit/82a11d2f4b7a2b747def6a5bfe1a52fd5a30a9ee))
- set to hardhat 2.10.0 and ethers to 5.6.9 ([9b50898](https://github.com/aave/aave-v3-core/commit/9b50898f6dbace6c2228e4bd41c081ff4afa7324))

### Bug Fixes

- dependencies ([f844a45](https://github.com/aave/aave-v3-core/commit/f844a4596b22fd16ec2516fa1e72c6d223481710))
- load market test data correctly, fix atoken/debt token names ([72d1264](https://github.com/aave/aave-v3-core/commit/72d1264d1e6a285663828e37c52b9078525ca291))
- remove npm ci cache, bump gas reporter to fixed version 1.0.8 set ethers to fixed version 5.6.1 ([bbb2dfd](https://github.com/aave/aave-v3-core/commit/bbb2dfde6c754c4e0552e78f43c5c09a6474805f))
- upgrade periphery and deploy library to latest version ([902b48a](https://github.com/aave/aave-v3-core/commit/902b48aca1b53435fb302c0ba462c79beb1b57b8))
- use ethers 5.5.3 to prevent different @ethersproject/bignumber version ([5411930](https://github.com/aave/aave-v3-core/commit/541193012dfd3c7ee1cb0dc2dfc11db091876145))

### Miscellaneous Chores

- release 1.16.2 ([32901d1](https://github.com/aave/aave-v3-core/commit/32901d1e541b38e2273ae896c43323a02a2ed744))

### [1.16.1](https://www.github.com/aave/aave-v3-core/compare/v1.16.0...v1.16.1) (2022-04-05)

### Bug Fixes

- add comment for undocumentted parameter ([ba0e4ee](https://www.github.com/aave/aave-v3-core/commit/ba0e4ee72cc3060fc55294aeac0a7bbb9caf3087))
- Add helpers for proxy contracts (eip1967 slots) ([d82be43](https://www.github.com/aave/aave-v3-core/commit/d82be4350d1778c269b5f7d875bdf5025fcff0bd))
- Fix ts type of contract in tests ([12373ca](https://www.github.com/aave/aave-v3-core/commit/12373ca09d4e2cec32546dc87285d516c1d9f261))

## [1.16.0](https://www.github.com/aave/aave-v3-core/compare/v1.15.0...v1.16.0) (2022-03-15)

### Features

- updated price oracle sentinel interface ([f6b71f5](https://www.github.com/aave/aave-v3-core/commit/f6b71f508db2a9a6f966bc6d606198eebaaefbb9))

### Bug Fixes

- Fix docstrings of ISequencerOracle ([4391fd4](https://www.github.com/aave/aave-v3-core/commit/4391fd4c8f29fb81c966fd4510df89d0acdb8404))

## [1.15.0](https://www.github.com/aave/aave-v3-core/compare/v1.14.2...v1.15.0) (2022-03-04)

### Features

- add owner constructor parameter to contracts that inherits Ownable to support CREATE2 factory deployment ([b6cc245](https://www.github.com/aave/aave-v3-core/commit/b6cc245b7e78e7d27b7af3e045cceff58e477231))
- bump beta deploy package ([fbcf885](https://www.github.com/aave/aave-v3-core/commit/fbcf885269285b36b54cc741da7922d257255a44))
- refactored executeLiquidationCall function ([63e43ef](https://www.github.com/aave/aave-v3-core/commit/63e43eff1c36a8d7e96c86f9655e7af31fe7e81f))

### Bug Fixes

- add owner parameters to test suites ([6e96821](https://www.github.com/aave/aave-v3-core/commit/6e968212d088ab6c9911b85dc454e4b989db7d6f))
- Move reservesData param to the beginning of the param list ([0872cb4](https://www.github.com/aave/aave-v3-core/commit/0872cb4c30e79566f09a0320168d0bd1f5111a29))

### [1.14.2](https://www.github.com/aave/aave-v3-core/compare/v1.14.1...v1.14.2) (2022-03-02)

### Bug Fixes

- Replace `...PriceAddress` with `...PriceSource` ([10a8667](https://www.github.com/aave/aave-v3-core/commit/10a8667b08c2e3daa1d87b1a7348154081c3903d))
- Use `EModeLogic::isInEModeCategory` in `executeLiquidationCall` ([28f72fe](https://www.github.com/aave/aave-v3-core/commit/28f72fe82496044d23d4ea3f20298cec9918404b))

### [1.14.1](https://www.github.com/aave/aave-v3-core/compare/v1.14.0...v1.14.1) (2022-03-01)

### Bug Fixes

- Improve consistency of function naming of CalldataLogic ([1a5517d](https://www.github.com/aave/aave-v3-core/commit/1a5517d28b531d673e6ac81e59b552d1280e7d7e))
- Use memory instead of storage in `getLiquidationBonus` ([deccf52](https://www.github.com/aave/aave-v3-core/commit/deccf529ad8d300f8a0e768cc8d9bf6d48b7699f))

## [1.14.0](https://www.github.com/aave/aave-v3-core/compare/v1.13.1...v1.14.0) (2022-02-21)

### Features

- clean dependencies and upgrade child dependencies ([7ca97ca](https://www.github.com/aave/aave-v3-core/commit/7ca97ca4cd6e9aad74583a4a1ca5ebe3fae44c64))
- removed obsolete files for the certora tools ([d7e0e7c](https://www.github.com/aave/aave-v3-core/commit/d7e0e7c37c0741b54afde3354d0be4a619fafde5))

### Bug Fixes

- `reserves` renamed to `reservesData` or `reservesList` when fitting ([3a6b928](https://www.github.com/aave/aave-v3-core/commit/3a6b928abb9ad6920582064704673e84f4b543dd))
- Cleanup naming and remove duplicate tests ([6ed5891](https://www.github.com/aave/aave-v3-core/commit/6ed589163ae30b235206dc571b98091f08dc5bad))
- Fix typo in package contributors tag ([14a3b6d](https://www.github.com/aave/aave-v3-core/commit/14a3b6d503d71d080c145e7fbcdacbe68b2b11ba))
- Moved getUserAccountData logic to PoolLogic to minimize contract size ([b07bdab](https://www.github.com/aave/aave-v3-core/commit/b07bdab921705788f38e130f75a10a1444cbc6b1))
- Natspec for `IPool::swapBorrowRateMode()` ([5b016fc](https://www.github.com/aave/aave-v3-core/commit/5b016fc699ef816204cf1e0fe1178285f53cf83c))
- Place interface extension declaration at the end ([19c015e](https://www.github.com/aave/aave-v3-core/commit/19c015ec76efc229da6b59fafa9f640d56e7f157))
- Rename reserves to reservesList ([4b0af0b](https://www.github.com/aave/aave-v3-core/commit/4b0af0bbb200e76dfedc7adf923c80e74fae37cb))
- Revert dependencies to last working state ([ff4d987](https://www.github.com/aave/aave-v3-core/commit/ff4d98765cea9b07f3e61f5b5e0efc472a06addc))
- typo in tech paper ([d6b9cbc](https://www.github.com/aave/aave-v3-core/commit/d6b9cbc55adf88c4cb7dd6df11941167fd119b3d))
- Update contributors etc in package.json ([60fc967](https://www.github.com/aave/aave-v3-core/commit/60fc967fa8037bb856c0c39c1e433fabafab255c))

helpers/types.ts
import {BigNumber, BigNumberish} from 'ethers';

export interface SymbolMap<T> {
  [symbol: string]: T;
}

export type eNetwork = eEthereumNetwork;

export enum eEthereumNetwork {
  goerli = 'goerli',
  main = 'main',
  coverage = 'coverage',
  hardhat = 'hardhat',
  polygon = 'polygon',
}

export enum eContractid {
  Example = 'Example',
  PoolAddressesProvider = 'PoolAddressesProvider',
  MintableERC20 = 'MintableERC20',
  MintableDelegationERC20 = 'MintableDelegationERC20',
  PoolAddressesProviderRegistry = 'PoolAddressesProviderRegistry',
  ACLManager = 'ACLManager',
  PoolParametersProvider = 'PoolParametersProvider',
  PoolConfigurator = 'PoolConfigurator',
  ValidationLogic = 'ValidationLogic',
  ReserveLogic = 'ReserveLogic',
  GenericLogic = 'GenericLogic',
  SupplyLogic = 'SupplyLogic',
  BorrowLogic = 'BorrowLogic',
  FlashLoanLogic = 'FlashLoanLogic',
  LiquidationLogic = 'LiquidationLogic',
  BridgeLogic = 'BridgeLogic',
  EModeLogic = 'EModeLogic',
  ConfiguratorLogic = 'ConfiguratorLogic',
  Pool = 'Pool',
  PriceOracle = 'PriceOracle',
  Proxy = 'Proxy',
  MockAggregator = 'MockAggregator',
  AaveOracle = 'AaveOracle',
  DefaultReserveInterestRateStrategy = 'DefaultReserveInterestRateStrategy',
  InitializableImmutableAdminUpgradeabilityProxy = 'InitializableImmutableAdminUpgradeabilityProxy',
  MockFlashLoanReceiver = 'MockFlashLoanReceiver',
  AToken = 'AToken',
  MockAToken = 'MockAToken',
  DelegationAwareAToken = 'DelegationAwareAToken',
  MockStableDebtToken = 'MockStableDebtToken',
  MockVariableDebtToken = 'MockVariableDebtToken',
  AaveProtocolDataProvider = 'AaveProtocolDataProvider',
  IERC20Detailed = 'IERC20Detailed',
  StableDebtToken = 'StableDebtToken',
  VariableDebtToken = 'VariableDebtToken',
  FeeProvider = 'FeeProvider',
  TokenDistributor = 'TokenDistributor',
  ReservesSetupHelper = 'ReservesSetupHelper',
  WETH = 'WETH',
  WETHMocked = 'WETHMocked',
  PoolImpl = 'PoolImpl',
  PoolConfiguratorImpl = 'PoolConfiguratorImpl',
  MockIncentivesController = 'MockIncentivesController',
  MockReserveConfiguration = 'MockReserveConfiguration',
  MockPool = 'MockPool',
  MockInitializableImple = 'MockInitializableImple',
  MockInitializableImpleV2 = 'MockInitializableImpleV2',
  MockInitializableFromConstructorImple = 'MockInitializableFromConstructorImple',
  MockReentrantInitializableImple = 'MockReentrantInitializableImple',
}

/*
 * Error messages
 */
export enum ProtocolErrors {
  CALLER_NOT_POOL_ADMIN = '1', // 'The caller of the function is not a pool admin'
  CALLER_NOT_EMERGENCY_ADMIN = '2', // 'The caller of the function is not an emergency admin'
  CALLER_NOT_POOL_OR_EMERGENCY_ADMIN = '3', // 'The caller of the function is not a pool or emergency admin'
  CALLER_NOT_RISK_OR_POOL_ADMIN = '4', // 'The caller of the function is not a risk or pool admin'
  CALLER_NOT_ASSET_LISTING_OR_POOL_ADMIN = '5', // 'The caller of the function is not an asset listing or pool admin'
  CALLER_NOT_BRIDGE = '6', // 'The caller of the function is not a bridge'
  ADDRESSES_PROVIDER_NOT_REGISTERED = '7', // 'Pool addresses provider is not registered'
  INVALID_ADDRESSES_PROVIDER_ID = '8', // 'Invalid id for the pool addresses provider'
  NOT_CONTRACT = '9', // 'Address is not a contract'
  CALLER_NOT_POOL_CONFIGURATOR = '10', // 'The caller of the function is not the pool configurator'
  CALLER_NOT_ATOKEN = '11', // 'The caller of the function is not an AToken'
  INVALID_ADDRESSES_PROVIDER = '12', // 'The address of the pool addresses provider is invalid'
  INVALID_FLASHLOAN_EXECUTOR_RETURN = '13', // 'Invalid return value of the flashloan executor function'
  RESERVE_ALREADY_ADDED = '14', // 'Reserve has already been added to reserve list'
  NO_MORE_RESERVES_ALLOWED = '15', // 'Maximum amount of reserves in the pool reached'
  EMODE_CATEGORY_RESERVED = '16', // 'Zero eMode category is reserved for volatile heterogeneous assets'
  INVALID_EMODE_CATEGORY_ASSIGNMENT = '17', // 'Invalid eMode category assignment to asset'
  RESERVE_LIQUIDITY_NOT_ZERO = '18', // 'The liquidity of the reserve needs to be 0'
  FLASHLOAN_PREMIUM_INVALID = '19', // 'Invalid flashloan premium'
  INVALID_RESERVE_PARAMS = '20', // 'Invalid risk parameters for the reserve'
  INVALID_EMODE_CATEGORY_PARAMS = '21', // 'Invalid risk parameters for the eMode category'
  BRIDGE_PROTOCOL_FEE_INVALID = '22', // 'Invalid bridge protocol fee'
  CALLER_MUST_BE_POOL = '23', // 'The caller of this function must be a pool'
  INVALID_MINT_AMOUNT = '24', // 'Invalid amount to mint'
  INVALID_BURN_AMOUNT = '25', // 'Invalid amount to burn'
  INVALID_AMOUNT = '26', // 'Amount must be greater than 0'
  RESERVE_INACTIVE = '27', // 'Action requires an active reserve'
  RESERVE_FROZEN = '28', // 'Action cannot be performed because the reserve is frozen'
  RESERVE_PAUSED = '29', // 'Action cannot be performed because the reserve is paused'
  BORROWING_NOT_ENABLED = '30', // 'Borrowing is not enabled'
  STABLE_BORROWING_NOT_ENABLED = '31', // 'Stable borrowing is not enabled'
  NOT_ENOUGH_AVAILABLE_USER_BALANCE = '32', // 'User cannot withdraw more than the available balance'
  INVALID_INTEREST_RATE_MODE_SELECTED = '33', // 'Invalid interest rate mode selected'
  COLLATERAL_BALANCE_IS_ZERO = '34', // 'The collateral balance is 0'
  HEALTH_FACTOR_LOWER_THAN_LIQUIDATION_THRESHOLD = '35', // 'Health factor is lesser than the liquidation threshold'
  COLLATERAL_CANNOT_COVER_NEW_BORROW = '36', // 'There is not enough collateral to cover a new borrow'
  COLLATERAL_SAME_AS_BORROWING_CURRENCY = '37', // 'Collateral is (mostly) the same currency that is being borrowed'
  AMOUNT_BIGGER_THAN_MAX_LOAN_SIZE_STABLE = '38', // 'The requested amount is greater than the max loan size in stable rate mode'
  NO_DEBT_OF_SELECTED_TYPE = '39', // 'For repayment of a specific type of debt, the user needs to have debt that type'
  NO_EXPLICIT_AMOUNT_TO_REPAY_ON_BEHALF = '40', // 'To repay on behalf of a user an explicit amount to repay is needed'
  NO_OUTSTANDING_STABLE_DEBT = '41', // 'User does not have outstanding stable rate debt on this reserve'
  NO_OUTSTANDING_VARIABLE_DEBT = '42', // 'User does not have outstanding variable rate debt on this reserve'
  UNDERLYING_BALANCE_ZERO = '43', // 'The underlying balance needs to be greater than 0'
  INTEREST_RATE_REBALANCE_CONDITIONS_NOT_MET = '44', // 'Interest rate rebalance conditions were not met'
  HEALTH_FACTOR_NOT_BELOW_THRESHOLD = '45', // 'Health factor is not below the threshold'
  COLLATERAL_CANNOT_BE_LIQUIDATED = '46', // 'The collateral chosen cannot be liquidated'
  SPECIFIED_CURRENCY_NOT_BORROWED_BY_USER = '47', // 'User did not borrow the specified currency'
  SAME_BLOCK_BORROW_REPAY = '48', // 'Borrow and repay in same block is not allowed'
  INCONSISTENT_FLASHLOAN_PARAMS = '49', // 'Inconsistent flashloan parameters'
  BORROW_CAP_EXCEEDED = '50', // 'Borrow cap is exceeded'
  SUPPLY_CAP_EXCEEDED = '51', // 'Supply cap is exceeded'
  UNBACKED_MINT_CAP_EXCEEDED = '52', // 'Unbacked mint cap is exceeded'
  DEBT_CEILING_EXCEEDED = '53', // 'Debt ceiling is exceeded'
  ATOKEN_SUPPLY_NOT_ZERO = '54', // 'AToken supply is not zero'
  STABLE_DEBT_NOT_ZERO = '55', // 'Stable debt supply is not zero'
  VARIABLE_DEBT_SUPPLY_NOT_ZERO = '56', // 'Variable debt supply is not zero'
  LTV_VALIDATION_FAILED = '57', // 'Ltv validation failed'
  INCONSISTENT_EMODE_CATEGORY = '58', // 'Inconsistent eMode category'
  PRICE_ORACLE_SENTINEL_CHECK_FAILED = '59', // 'Price oracle sentinel validation failed'
  ASSET_NOT_BORROWABLE_IN_ISOLATION = '60', // 'Asset is not borrowable in isolation mode'
  RESERVE_ALREADY_INITIALIZED = '61', // 'Reserve has already been initialized'
  USER_IN_ISOLATION_MODE = '62', // 'User is in isolation mode'
  INVALID_LTV = '63', // 'Invalid ltv parameter for the reserve'
  INVALID_LIQ_THRESHOLD = '64', // 'Invalid liquidity threshold parameter for the reserve'
  INVALID_LIQ_BONUS = '65', // 'Invalid liquidity bonus parameter for the reserve'
  INVALID_DECIMALS = '66', // 'Invalid decimals parameter of the underlying asset of the reserve'
  INVALID_RESERVE_FACTOR = '67', // 'Invalid reserve factor parameter for the reserve'
  INVALID_BORROW_CAP = '68', // 'Invalid borrow cap for the reserve'
  INVALID_SUPPLY_CAP = '69', // 'Invalid supply cap for the reserve'
  INVALID_LIQUIDATION_PROTOCOL_FEE = '70', // 'Invalid liquidation protocol fee for the reserve'
  INVALID_EMODE_CATEGORY = '71', // 'Invalid eMode category for the reserve'
  INVALID_UNBACKED_MINT_CAP = '72', // 'Invalid unbacked mint cap for the reserve'
  INVALID_DEBT_CEILING = '73', // 'Invalid debt ceiling for the reserve
  INVALID_RESERVE_INDEX = '74', // 'Invalid reserve index'
  ACL_ADMIN_CANNOT_BE_ZERO = '75', // 'ACL admin cannot be set to the zero address'
  INCONSISTENT_PARAMS_LENGTH = '76', // 'Array parameters that should be equal length are not'
  ZERO_ADDRESS_NOT_VALID = '77', // 'Zero address not valid'
  INVALID_EXPIRATION = '78', // 'Invalid expiration'
  INVALID_SIGNATURE = '79', // 'Invalid signature'
  OPERATION_NOT_SUPPORTED = '80', // 'Operation not supported'
  DEBT_CEILING_NOT_ZERO = '81', // 'Debt ceiling is not zero'
  ASSET_NOT_LISTED = '82', // 'Asset is not listed'
  INVALID_OPTIMAL_USAGE_RATIO = '83', // 'Invalid optimal usage ratio'
  INVALID_OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO = '84', // 'Invalid optimal stable to total debt ratio'
  UNDERLYING_CANNOT_BE_RESCUED = '85', // 'The underlying asset cannot be rescued'
  ADDRESSES_PROVIDER_ALREADY_ADDED = '86', // 'Reserve has already been added to reserve list'
  POOL_ADDRESSES_DO_NOT_MATCH = '87', // 'The token implementation pool address and the pool address provided by the initializing pool do not match'
  STABLE_BORROWING_ENABLED = '88', // 'Stable borrowing is enabled'
  SILOED_BORROWING_VIOLATION = '89', // user is trying to violate the siloed borrowing rule
  RESERVE_DEBT_NOT_ZERO = '90', // the total debt of the reserve needs to be 0
  // SafeCast
  SAFECAST_UINT128_OVERFLOW = "SafeCast: value doesn't fit in 128 bits",

  // Ownable
  OWNABLE_ONLY_OWNER = 'Ownable: caller is not the owner',

  // ERC20
  ERC20_TRANSFER_AMOUNT_EXCEEDS_BALANCE = 'ERC20: transfer amount exceeds balance',

  // old

  INVALID_FROM_BALANCE_AFTER_TRANSFER = 'Invalid from balance after transfer',
  INVALID_TO_BALANCE_AFTER_TRANSFER = 'Invalid from balance after transfer',
  INVALID_HF = 'Invalid health factor',
}

export type tEthereumAddress = string;
export type tStringTokenBigUnits = string; // 1 ETH, or 10e6 USDC or 10e18 DAI
export type tBigNumberTokenBigUnits = BigNumber;
export type tStringTokenSmallUnits = string; // 1 wei, or 1 basic unit of USDC, or 1 basic unit of DAI
export type tBigNumberTokenSmallUnits = BigNumber;

export interface iAssetCommon<T> {
  [key: string]: T;
}
export interface iAssetBase<T> {
  WETH: T;
  DAI: T;
  TUSD: T;
  USDC: T;
  USDT: T;
  SUSD: T;
  AAVE: T;
  BAT: T;
  MKR: T;
  LINK: T;
  KNC: T;
  WBTC: T;
  MANA: T;
  ZRX: T;
  SNX: T;
  BUSD: T;
  YFI: T;
  UNI: T;
  USD: T;
  REN: T;
  ENJ: T;
  UniDAIWETH: T;
  UniWBTCWETH: T;
  UniAAVEWETH: T;
  UniBATWETH: T;
  UniDAIUSDC: T;
  UniCRVWETH: T;
  UniLINKWETH: T;
  UniMKRWETH: T;
  UniRENWETH: T;
  UniSNXWETH: T;
  UniUNIWETH: T;
  UniUSDCWETH: T;
  UniWBTCUSDC: T;
  UniYFIWETH: T;
  BptWBTCWETH: T;
  BptBALWETH: T;
  WMATIC: T;
  STAKE: T;
  xSUSHI: T;
}

export type iAssetsWithoutETH<T> = Omit<iAssetBase<T>, 'ETH'>;

export type iAssetsWithoutUSD<T> = Omit<iAssetBase<T>, 'USD'>;

export type iAavePoolAssets<T> = Pick<
  iAssetsWithoutUSD<T>,
  | 'DAI'
  | 'TUSD'
  | 'USDC'
  | 'USDT'
  | 'SUSD'
  | 'AAVE'
  | 'BAT'
  | 'MKR'
  | 'LINK'
  | 'KNC'
  | 'WBTC'
  | 'MANA'
  | 'ZRX'
  | 'SNX'
  | 'BUSD'
  | 'WETH'
  | 'YFI'
  | 'UNI'
  | 'REN'
  | 'ENJ'
  | 'xSUSHI'
>;

export type iMultiPoolsAssets<T> = iAssetCommon<T> | iAavePoolAssets<T>;

export type iAssetAggregatorBase<T> = iAssetsWithoutETH<T>;

export enum TokenContractId {
  DAI = 'DAI',
  AAVE = 'AAVE',
  TUSD = 'TUSD',
  BAT = 'BAT',
  WETH = 'WETH',
  USDC = 'USDC',
  USDT = 'USDT',
  SUSD = 'SUSD',
  ZRX = 'ZRX',
  MKR = 'MKR',
  WBTC = 'WBTC',
  LINK = 'LINK',
  KNC = 'KNC',
  MANA = 'MANA',
  REN = 'REN',
  SNX = 'SNX',
  BUSD = 'BUSD',
  USD = 'USD',
  YFI = 'YFI',
  UNI = 'UNI',
  ENJ = 'ENJ',
  UniDAIWETH = 'UniDAIWETH',
  UniWBTCWETH = 'UniWBTCWETH',
  UniAAVEWETH = 'UniAAVEWETH',
  UniBATWETH = 'UniBATWETH',
  UniDAIUSDC = 'UniDAIUSDC',
  UniCRVWETH = 'UniCRVWETH',
  UniLINKWETH = 'UniLINKWETH',
  UniMKRWETH = 'UniMKRWETH',
  UniRENWETH = 'UniRENWETH',
  UniSNXWETH = 'UniSNXWETH',
  UniUNIWETH = 'UniUNIWETH',
  UniUSDCWETH = 'UniUSDCWETH',
  UniWBTCUSDC = 'UniWBTCUSDC',
  UniYFIWETH = 'UniYFIWETH',
  BptWBTCWETH = 'BptWBTCWETH',
  BptBALWETH = 'BptBALWETH',
  WMATIC = 'WMATIC',
  STAKE = 'STAKE',
  xSUSHI = 'xSUSHI',
}

export interface IReserveParams extends IReserveBorrowParams, IReserveCollateralParams {
  aTokenImpl: eContractid;
  reserveFactor: string;
  supplyCap: string;
  strategy: IInterestRateStrategyParams;
}

export interface IInterestRateStrategyParams {
  name: string;
  optimalUsageRatio: string;
  baseVariableBorrowRate: string;
  variableRateSlope1: string;
  variableRateSlope2: string;
  stableRateSlope1: string;
  stableRateSlope2: string;
  baseStableRateOffset: string;
  stableRateExcessOffset: string;
  optimalStableToTotalDebtRatio: string;
}

export interface IReserveBorrowParams {
  borrowingEnabled: boolean;
  stableBorrowRateEnabled: boolean;
  reserveDecimals: string;
  borrowCap: string;
  flashLoanEnabled?: boolean;
  debtCeiling?: string;
  borrowableIsolation?: boolean;
}

export interface IReserveCollateralParams {
  baseLTVAsCollateral: string;
  liquidationThreshold: string;
  liquidationBonus: string;
  liquidationProtocolFee?: string;
}
export interface IMarketRates {
  borrowRate: string;
}

export type iParamsPerNetwork<T> = iEthereumParamsPerNetwork<T>;

export interface iParamsPerNetworkAll<T> extends iEthereumParamsPerNetwork<T> {}

export interface iEthereumParamsPerNetwork<T> {
  [eEthereumNetwork.coverage]: T;
  [eEthereumNetwork.goerli]: T;
  [eEthereumNetwork.main]: T;
  [eEthereumNetwork.hardhat]: T;
  [eEthereumNetwork.polygon]: T;
}

export enum RateMode {
  None = '0',
  Stable = '1',
  Variable = '2',
}

export interface IProtocolGlobalConfig {
  TokenDistributorPercentageBase: string;
  MockUsdPriceInWei: string;
  UsdAddress: tEthereumAddress;
  NilAddress: tEthereumAddress;
  OneAddress: tEthereumAddress;
  AaveReferral: string;
}

export interface IMocksConfig {
  AllAssetsInitialPrices: iAssetBase<string>;
}

export interface IRate {
  borrowRate: string;
}

export interface ICommonConfiguration {
  MarketId?: string;
  ATokenNamePrefix?: string;
  StableDebtTokenNamePrefix?: string;
  VariableDebtTokenNamePrefix?: string;
  SymbolPrefix?: string;
  OracleQuoteCurrencyAddress?: tEthereumAddress;
  OracleQuoteCurrency?: string;
  OracleQuoteUnit?: string;
  WrappedNativeTokenSymbol?: string;
  ProviderId?: number;
  ProtocolGlobalParams?: IProtocolGlobalConfig;
  Mocks?: IMocksConfig;
  ProviderRegistry?: tEthereumAddress | undefined;
  ProviderRegistryOwner?: tEthereumAddress | undefined;
  PoolConfigurator?: tEthereumAddress | undefined;
  Pool?: tEthereumAddress | undefined;
  TokenDistributor?: tEthereumAddress | undefined;
  AaveOracle?: tEthereumAddress | undefined;
  FallbackOracle?: tEthereumAddress | undefined | iParamsPerNetwork<tEthereumAddress>;
  ChainlinkAggregator?: tEthereumAddress | undefined | iParamsPerNetwork<ITokenAddress>;
  PoolAdmin?: tEthereumAddress | undefined;
  PoolAdminIndex?: number;
  EmergencyAdmin?: tEthereumAddress | undefined;
  EmergencyAdminIndex?: number;
  ReserveAssets?:
    | SymbolMap<tEthereumAddress>
    | SymbolMap<undefined>
    | iParamsPerNetwork<SymbolMap<tEthereumAddress>>;
  ReservesConfig?: iMultiPoolsAssets<IReserveParams>;
  ATokenDomainSeparator?: string;
  WETH?: tEthereumAddress | undefined;
  WrappedNativeToken?: tEthereumAddress | undefined;
  ReserveFactorTreasuryAddress?: tEthereumAddress | iParamsPerNetwork<tEthereumAddress>;
  IncentivesController?: tEthereumAddress | undefined;
  IncentivesConfig?: IncentivesConfig;
  EModes?: SymbolMap<EMode>;
  L2PoolEnabled?: iParamsPerNetwork<boolean>;
  ParaswapRegistry?: iParamsPerNetwork<tEthereumAddress>;
  FlashLoanPremiums?: {
    total: number;
    protocol: number;
  };
}

export interface IncentivesConfig {
  enabled: iParamsPerNetwork<boolean>;
  rewards: iParamsPerNetwork<SymbolMap<tEthereumAddress>>;
  rewardsOracle: iParamsPerNetwork<SymbolMap<tEthereumAddress>>;
  incentivesInput: iParamsPerNetwork<RewardsConfigInput[]>;
}

export interface EMode {
  id: string;
  ltv: string;
  liquidationThreshold: string;
  liquidationBonus: string;
  label: string;
  oracleId?: string;
  assets: string[];
}

export declare enum AssetType {
  AToken = 0,
  VariableDebtToken = 1,
  StableDebtToken = 2,
}

export interface RewardsConfigInput {
  emissionPerSecond: BigNumberish;
  duration: number;
  asset: string;
  assetType: AssetType;
  reward: string;
  rewardOracle: string;
  transferStrategy: TransferStrategy;
  transferStrategyParams: string;
}

export declare enum TransferStrategy {
  PullRewardsStrategy = 0,
  StakedRewardsStrategy = 1,
}

export interface ITokenAddress {
  [token: string]: tEthereumAddress;
}

export interface IAaveConfiguration extends ICommonConfiguration {
  ReservesConfig: iMultiPoolsAssets<IReserveParams>;
  StkAaveProxy?: iParamsPerNetwork<tEthereumAddress>;
}

export type PoolConfiguration = ICommonConfiguration | IAaveConfiguration;
